File: 0003_longest_substring_without_repeating_characters/sliding_window/go/solution.go
package longest_substring_without_repeating_characters

func lengthOfLongestSubstring(s string) int {
	lastSeen := make(map[rune]int)
	left := 0
	best := 0

	for right, ch := range []rune(s) {
		if previous, exists := lastSeen[ch]; exists && previous >= left {
			left = previous + 1
		}
		lastSeen[ch] = right
		if length := right - left + 1; length > best {
			best = length
		}
	}

	return best
}

File: 0003_longest_substring_without_repeating_characters/sliding_window/kotlin/solution.kt
fun lengthOfLongestSubstring(s: String): Int {
    val lastSeen = mutableMapOf<Char, Int>()
    var left = 0
    var best = 0

    for ((right, ch) in s.withIndex()) {
        val previous = lastSeen[ch]
        if (previous != null && previous >= left) {
            left = previous + 1
        }
        lastSeen[ch] = right
        best = maxOf(best, right - left + 1)
    }

    return best
}
